### DISCUSSION TITLE
Places to learn Pyglet and Ren'Py (With an IQ < 70)

### DISCUSSION THREAD

**Discussion by ghost** [community/community#102405] [Programming Help] [answered]
## Pyglet:
* Lighting (Daylight system, maps and more realism)
* Audio 
* Ray casting (Create 3D elements in Pyglet)
* Animations(Dying, traps, combat system, shooting, reloading, etc)
* Different story lines based on user choices
*  Player upgrades
 * NPCs
 * Water physics
 * Sword physics
 * Bow(weapon) physics
 * Gun physics
 * Gore physics
* Reverse Bear Claw Trap physics
 ## Ren'py: 
 * Story
 * User choices.
Story:

You are a detective who has been investigating the Jigsaw Killer, a notorious serial killer who traps his victims in elaborate and deadly games. You have been following a lead to an abandoned warehouse, where you hope to find some clues or evidence. However, as soon as you enter the warehouse, you hear a loud noise and the door slams shut behind you. You realize that you have fallen into a trap, and that you are not alone. There are other people in the warehouse, who have also been kidnapped by the Jigsaw Killer. You recognize some of them as criminals, suspects, or witnesses that you have encountered during your investigation. You also see a TV screen, where a puppet wearing a red cloak and a white mask appears. The puppet introduces itself as the Jigsaw Killer, and tells you that you and the others have been chosen to play a game. The game consists of several rooms, each containing a different trap. The traps are designed to test your morality, intelligence, and will to live. The puppet explains that you have two options: you can either work together with the others to escape the warehouse, or you can betray them and try to save yourself. The puppet warns you that the game has a time limit, and that if you fail to escape or survive, you will die. The puppet also tells you that each of you has a device attached to your neck, which will explode if you try to remove it or tamper with it. The puppet then says “Let the game begin”, and the TV screen turns off. You look around and see a door with a sign that says “Room 1”. You have to make a choice: will you cooperate or compete with the others? Will you follow the rules or break them? Will you live or die?

Backstory:

You: You are a detective who has been obsessed with catching the Jigsaw Killer. You have dedicated your life to solving the case, but you have also neglected your family, friends, and health. You have a reputation for being ruthless, arrogant, and reckless. You have broken the law and violated the rights of many suspects and witnesses in order to get information. You have also developed a drug addiction, which has impaired your judgment and performance. You have a wife and a daughter, who have left you because of your obsession. You have a partner, who is loyal and supportive, but also worried about you. You have a boss, who is frustrated and angry with you, and who has threatened to fire you or arrest you if you don’t stop your illegal actions. You have a motive to escape the warehouse and catch the Jigsaw Killer, but you also have to face the consequences of your actions and the mistakes of your past.
Alice: Alice is a journalist who has been covering the Jigsaw Killer case. She is ambitious, curious, and fearless. She has a knack for finding and exposing the truth, but she also has a tendency to be intrusive, manipulative, and sensationalist. She has hacked into police databases, bribed witnesses, and trespassed crime scenes in order to get exclusive stories. She has also published some articles that have endangered the lives of some victims or interfered with the investigation. She has a rival, who is another journalist who works for a competing newspaper. She has a source, who is a mysterious informant who has given her some tips and clues about the Jigsaw Killer. She has a motive to escape the warehouse and expose the Jigsaw Killer, but she also has to deal with the risks and ethics of her profession and the secrets of her source.
Bob: Bob is a drug dealer who has been involved in the Jigsaw Killer case. He is greedy, cunning, and violent. He has a network of contacts, customers, and suppliers, who he exploits and intimidates. He has also killed or injured some people who have crossed him or owed him money. He has a connection to the Jigsaw Killer, who he has supplied with some drugs and chemicals that have been used in some traps. He has a rival, who is another drug dealer who works for a different gang. He has a girlfriend, who is a drug addict who depends on him. He has a motive to escape the warehouse and continue his business, but he also has to face the wrath and revenge of the Jigsaw Killer and his rival.
Claire: Claire is a nurse who has been involved in the Jigsaw Killer case. She is compassionate, intelligent, and brave. She has a passion for helping and healing people, but she also has a sense of justice and morality. She has treated some victims of the Jigsaw Killer, who she has tried to save or comfort. She has also witnessed some atrocities and horrors that have traumatized her. She has a connection to the Jigsaw Killer, who she has met before as a patient. She has a husband, who is a doctor who works at the same hospital. She has a son, who is a teenager who is rebellious and troubled. She has a motive to escape the warehouse and reunite with her family, but she also has to cope with the pain and fear of her experience and the mystery of her connection.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by AleixMT** [comment, marked as answer]
Here are some of the learning materials that I found more useful regarding these two Python libraries:

# Learning materials

* [RenPy official guide](https://www.renpy.org/doc/html/quickstart.html): Quick start documentation of the RenPy library. 
* [Awesome renpy](https://github.com/methanoliver/awesome-renpy): Repository that stores a lot of resources to other repos and materials that you can use in RenPy. This is the **most** important resource since usually it is the gate to all the other resources. 
* [Pyglett guide in geeks for geeks](https://www.geeksforgeeks.org/introduction-to-pyglet-library-for-game-development-in-python/): Very nice guide to learn the basics of Pyglett in a practical sense. Very useful for people with <70 IQ 🤣 
* [Read the docs from Pyglett](https://pyglet.readthedocs.io/en/latest/): The official documentation for the Pyglett project. Exhaustive documentatioln where you should find almost everything about technical aspects of Pyglett.
* [RenPy Reddit community](https://www.reddit.com/r/RenPy/): This is really an awesome place to ask for help if something that you do not understand happens.

With this you have learning materials for months. You also have the possibility to ask for help in a dedicated community. 

Since you introduced the backstory of your characters and the elements that you want to work with, I created a small schema about how can you implement each aspect of your game. With this you have a small schema of all the different elements that compose your game and how can you combine them using the libraries that you asked (specially PyGlett):

# Game elements

### Lighting (Daylight system, maps, and more realism):

Pyglet provides a straightforward way to handle lighting effects by manipulating OpenGL shaders. You can create day and night cycles by adjusting the intensity of light sources.
Implementing realistic maps involves creating 3D models and rendering them in the game world. Pyglet's OpenGL support can help with this.

### Audio:

Pyglet has built-in support for audio. You can use it to play background music, sound effects, and implement dynamic audio based on in-game events.

### Ray Casting (Create 3D elements in Pyglet):

Ray casting involves simulating the path of light rays in a 3D environment. Pyglet, with its OpenGL support, can handle the rendering of 3D elements and facilitate ray casting.

### Animations (Dying, traps, combat system, shooting, reloading, etc.):

Pyglet supports sprite animations. You can create sprite sheets for characters and objects, then animate them based on in-game events.
Implementing a combat system involves handling collision detection and responding to user input. Pyglet's event handling can be used for this purpose.

### Different storylines based on user choices:

You can manage branching storylines by keeping track of user choices using variables. Based on these choices, you can direct the player to different story paths. As you already know, Ren'py is the right library to use for this.

### Player upgrades:

Implement a system to track the player's progress and assign upgrades accordingly. This could include improving skills, unlocking new abilities, or obtaining better equipment.

### NPCs:

Create non-playable characters using Pyglet's sprite and animation capabilities. Implement AI behaviors for NPCs to interact with the player and each other.
Water Physics, Sword Physics, Bow (Weapon) Physics, Gun Physics, Gore Physics, Reverse Bear Claw Trap 

### Physics:

Pyglet provides basic physics functionalities, but for more realistic simulations, you might need to implement a physics engine like Pygame's PyBox2D.

### Ren'py for Story and User Choices:

Use Ren'Py for managing the overarching story, dialogues, and choices. It's a powerful visual novel engine that can handle complex narrative structures.

I hope that with this information you have a good beginning point to start programming your game!

If this answer was useful to you please mark it as the correct answer. Thank you!

**Reply by ghost** [reply]
Can someone like me still make a complex game like Saw with realistic lighting?
